packages/grpc-native-core/index.d.ts

```diff
@@ -0,0 +1,1214 @@
+     */
+    addProtoService(service: Service, implementation: { [name: string]: handleCall }): void;
+
+    /**
+     * Binds the server to the given port, with SSL disabled if creds is an
+     * insecure credentials object
+     * @param port The port that the server should bind on, in the format
+     * "address:port"
+     * @param creds Server credential object to be used for SSL. Pass an
+     * insecure credentials object for an insecure port.
+     */
+    bind(port: string, creds: ServerCredentials): void;
+  }
+
+  /**
+   * An object that completely defines a service.
+   * @typedef {Object.<string, grpc~MethodDefinition>} grpc~ServiceDefinition
+   */
+  export interface ServiceDefinition {
+    [s: string]: MethodDefinition;
+  }
+
+  /**
+   * An object that completely defines a service method signature.
+   */
+  export interface MethodDefinition {
+    /**
+     * The method's URL path
+     */
+    path: string;
+    /**
+     * Indicates whether the method accepts a stream of requests
+     */
+    requestStream: boolean;
+    /**
+     *  Indicates whether the method returns a stream of responses
+     */
+   responseStream: boolean;
+   /**
+    * Serialization function for request values
+    */
+    requestSerialize: serialize;
+    /**
+     * Serialization function for response values
+     */
+    responseSerialize: serialize;
+    /**
+     * Deserialization function for request data
+     */
+    requestDeserialize: deserialize;
+    /**
+     * Deserialization function for repsonse data
+     */
+    responseDeserialize: deserialize;
+  }
+
+  type handleCall = handleUnaryCall | handleClientStreamingCall | handleServerStreamingCall | handleBidiStreamingCall;
+
+  /**
+   * User-provided method to handle unary requests on a server
+   */
+  type handleUnaryCall = (call: ServerUnaryCall, callback: sendUnaryData) => void;
+
+  export class ServerUnaryCall {
+    /**
+     * Indicates if the call has been cancelled
+     */
+    cancelled: boolean;
+
+    /**
+     * The request metadata from the client
+     */
+    metadata: Metadata;
+
+    /**
+     * The request message from the client
+     */
+    request: any;
+
+    /**
+     * An EventEmitter. Used for unary calls.
+     * @param call The call object associated with the request
+     * @param metadata The request metadata from the client
+     */
+    constructor(call: Call, metadata: Metadata);
+
+    /**
+     * Get the endpoint this call/stream is connected to.
+     * @return The URI of the endpoint
+     */
+    getPeer(): string;
+
+    /**
+     * Send the initial metadata for a writable stream.
+     * @param responseMetadata Metadata to send
+     */
+    sendMetadata(responseMetadata: Metadata): void;
+  }
+
+  /**
+   * User provided method to handle client streaming methods on the server.
+   */
+  type handleClientStreamingCall = (call: ServerReadableStream, callback: sendUnaryData) => void;
+
+  export class ServerReadableStream {
```

Comment:
In the actual API, this class inherits from a Readable stream in object mode. Could the fact that those APIs are not in the type definitions cause problems for users? This applies to all of the streaming call types on both the client and server side, as well as the `UnaryClientCall` type, which inherits from EventEmitter.

Reply:
I'm not sure about this one. I think we need to explicitly inherit from those types as well otherwise something like this would cause an error.

```typescript
let c: ServerUnaryCall;
c.on("data", () => {});
```

To make this possible we need the Node.js typings. Should I include them as a dev dependency?

Reply:
I'm OK with depending on the Node.js typings. Wouldn't we have to include them as a regular dependency, if this `index.d.ts` we're distributing in the release references it?

Also, it's not as simple as just inheriting from those types. Those types are explicitly written with the assumption that the stream is a byte stream, and it takes some intermediate type redefinitions to make them valid object stream types. You can see what I had to do for this in the other package here: https://github.com/grpc/grpc-node/blob/master/packages/grpc-js-core/src/object-stream.ts

Reply:
You shouldn't have to do as much as that, because you don't necessarily need strongly typed object streams that restrict inputs and outputs to the specific request and response types. But you at least need equivalents of the intermediate types to allow `any` in the inputs to `write` operations and the outputs from `read` operations.

Reply:
I'll look into the definitions and see what I can do with it.

Reply:
All right, I've looked into this. The tests allow for your requested API, so I think we are good for now.

Reply:
I don't think the changes you made here quite work. As I mentioned, the builtin stream types are written for byte streams. They handle buffers and strings, not arbitrary objects. You need to override the method definitions, like I did in that object stream TypeScript file.

Also, it looks like you didn't change all of the client stream types.

Reply:
When looking into the Node.js typings it seems to allow the `any` type for both `read` results ([stream.Readable](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts#L5030)) and `write` messages ([stream.Writeable](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts#L5116)). 

See this test file where it's possible to write an object: https://github.com/lunarway/grpc-ts-test/blob/master/index.ts#L42

Let me know if I've misunderstood something here?

Reply:
On the readable side, you can see that the `data` event restricts the values to `Buffer | string`: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts#L5067. The `Writable` type is actually fine.

Reply:
The [first overload](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/node/index.d.ts#L5065) of `on` allows for the `any` type when reading data events. 

```typescript
readableStream.on("data", (chunk: object) => {});
```

I've made a test case in this commit: https://github.com/lunarway/grpc-ts-test/commit/23a74d585e2701a993f0b874ad017152cbc41083